### ISSUE TITLE
Revisiting Issue 304

### ISSUE THREAD

**Issue by ruolin59** [eslint/eslint#10870]
**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

see https://github.com/eslint/eslint/issues/7385

**Why should this rule be included in ESLint (instead of a plugin)?**
Revisiting this issue because previous responders (incorrectly) state that 

> Sometimes an async function won't be able to reasonably handle an error, in which case the Promise will reject and the error can instead be handled by the caller of the function.

which is not the case. In the case that the async function cannot handle the error, it needs to be caught and thrown immediately to ensure that it bubbles up the call stack. Otherwise the error disappears. 

From https://blog.grossman.io/how-to-write-async-await-without-try-catch-blocks-in-javascript/:
> ... The problem with await is that it will silently exit your function if no try-catch block was provided for it. And you won't have a way to control it unless providing the catch clause.

Of course, one can also handle errors in the promise using `.catch()` on the promise, but one of the two must be present to ensure proper error handling.

**Comment by not-an-aardvark** [comment]
Hi, thanks for creating an issue.

> In the case that the async function cannot handle the error, it needs to be caught and thrown immediately to ensure that it bubbles up the call stack.

This is not correct. A thrown error from an `await`-ed Promise in an async function will automatically bubble up the callstack, without needing to explicitly catch/rethrow it.

> The problem with await is that it will silently exit your function if no try-catch block was provided for it.

This is technically correct, but only in the same way that using `throw` will "silently" exit your function if no try-catch block was provided for it. In other words, it will do an early exit from your function, but the error will still end up bubbling up the call stack.

For example, consider this code:

```js
async function foo() {
   await somethingThatMightFail();
}

async function bar() {
    await somethingElseThatMightFail();
}

async function baz() {
    try {
        await foo();
        await bar();
    } catch (err) {
        handleError(err);
    }
}

baz();
```

If `somethingThatMightFail()` or `somethingElseThatMightFail()` throw an error, then the error will be caught by the `baz` function, despite the fact that the `foo` and `bar` functions don't explicitly handle the error. So it doesn't seem like it would make sense to require `foo` and `bar` to include a try-catch.

**Comment by ruolin59** [comment]
Thanks for the clarification with code examples. Turns out I had a bug handling errors in the top most level of my code base and that was the reason the errors were slipping.
